internal/mesh/internal/controllers/gatewayproxy/controller.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package gatewayproxy
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/controller/dependency"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/gatewayproxy/builder"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/gatewayproxy/fetcher"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecarproxy/cache"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// ControllerName is the name for this controller. It's used for logging or status keys.
+const ControllerName = "consul.io/gateway-proxy-controller"
+
+// Controller is responsible for triggering reconciler for watched resources
+func Controller(cache *cache.Cache) *controller.Controller {
+	// TODO Add the host of other types we should watch
+	return controller.NewController(ControllerName, pbmesh.ProxyStateTemplateType).
+		WithWatch(pbcatalog.WorkloadType, dependency.ReplaceType(pbmesh.ProxyStateTemplateType)).
+		WithWatch(pbmesh.ComputedProxyConfigurationType, dependency.ReplaceType(pbmesh.ProxyStateTemplateType)).
+		WithReconciler(&reconciler{
+			cache: cache,
+		})
+}
+
+// reconciler is responsible for managing the ProxyStateTemplate for all
+// gateway types: mesh, api (future) and terminating (future).
+type reconciler struct {
+	cache *cache.Cache
+}
+
+// Reconcile is responsible for creating and updating the pbmesh.ProxyStateTemplate
+// for all gateway types. Since the ProxyStateTemplates managed here will always have
+// an owner reference pointing to the corresponding pbmesh.MeshGateway, deletion is
+// left to the garbage collector.
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("resource-id", req.ID)
```

Comment:
is it possible we'll ever be running this reconciler in more than one thread? if so I don't think this assignment is threadsafe and we might want to use a local var instead

Reply:
You are correct. Since virtually every v2 resource controller that exists today uses this pattern, I'm fine leaving it and knowing that any change in the future would have to update all the other controllers as well. Thoughts?

https://github.com/hashicorp/consul/blob/010bf533d11ac29a5632659cd2af2c5120a11f8e/internal/mesh/internal/controllers/sidecarproxy/controller.go#L111

https://github.com/hashicorp/consul/blob/010bf533d11ac29a5632659cd2af2c5120a11f8e/internal/mesh/internal/controllers/sidecarproxy/controller.go#L111

and on and on...

Reply:
All controllers today operate as a single thread. Each dependency mapper IIRC runs in its own goroutine though.